aws/resource_aws_storagegateway_cache.go

```diff
@@ -53,6 +53,32 @@ func resourceAwsStorageGatewayCacheCreate(d *schema.ResourceData, meta interface
 		return fmt.Errorf("error adding Storage Gateway cache: %s", err)
 	}
 
+	// Depending on the Storage Gateway software, it will sometimes relabel a local DiskId
+	// with a UUID if previously unlabeled, e.g.
+	//   Before conn.AddCache(): "DiskId": "/dev/xvdb",
+	//   After conn.AddCache(): "DiskId": "112764d7-7e83-42ce-9af3-d482985a31cc",
+	// This prevents us from successfully reading the disk after creation.
+	// Here we try to refresh the local disks to see if we can find a new DiskId.
+
+	listLocalDisksInput := &storagegateway.ListLocalDisksInput{
+		GatewayARN: aws.String(gatewayARN),
+	}
+
+	log.Printf("[DEBUG] Reading Storage Gateway Local Disk: %s", listLocalDisksInput)
+	output, err := conn.ListLocalDisks(listLocalDisksInput)
+	if err != nil {
+		return fmt.Errorf("error reading Storage Gateway Local Disk: %s", err)
```

Comment:
Should we be doing `return` here?

I'm lacking some context, so bare with me. We don't seem to record the results of `AddCache`, what is created? Do we need to track it?

If `conn.AddCache(input)` above is successful, Terraform has _created_ something. If this call errors for whatever reason, valid or not, and we return before calling `d.SetId`, Terraform quits and we've lost track of whatever we created. 

Again, lacking context, sorry 😰

Reply:
Sorry if my comment note wasn't clear enough. 😅 

We're basically allocating a disk for a specified usage (cache) within a gateway (e.g. EC2 instance). When allocating said disk, the gateway (sometimes, but annoyingly) changes the disk identifier. Without refreshing the disks to try and find the new disk identifier, Terraform cannot track the resource as the cache is now based on the new disk identifier and there's no way to map between the old and new identifier.

That does make a good point though that it should probably still call the `d.SetId()` with the potentially wrong identifier before this API call just in case it does error and the disk identifier potentially doesn't change. I'll push up a commit momentarily.